modules/standard/Lists.chpl

```diff
@@ -0,0 +1,868 @@
+        ref item = _getRef(i);
+        _destroy(item);
+      }
+
+      if _blocks != nil then
+        for i in 1.._blocks.capacity do {
+          ref block = _blocks[i];
+          if block == nil then
+            continue;
+          _totalCapacity -= block.capacity;
+          delete block;
+          block = nil;
+        }
+
+      if _blocks != nil then
+        delete _blocks;
+
+      _sanity(_totalCapacity == 0);
+      _blocks = nil;
+      _size = 0;
+
+      _leave();
+    }
+
+    /*
+      Return a zero-based index into this List of the first item whose value
+      is equal to x.
+
+      :arg x: An element to search for.
+      :arg start: The start index to start searching from.
+      :arg end: The end index to stop searching at.
+
+      :return: The index of the element to search for.
+
+      :throws IllegalArgumentError: If the given element cannot be found.
+    */
+    proc indexOf(x: eltType, start: int=1, end: int=size): int throws {
+      _enter();
+
+      try _boundsCheckLeaveOnThrow(start);
+      try _boundsCheckLeaveOnThrow(end);
+
+      for i in start..end do
+        if x == _getRef(i) then {
+          _leave();
+          return i;
+        }
+
+      _leave();
+
+      const msg = "No such element in List: " + x:string;
+      throw new owned
+        IllegalArgumentError(msg);
+
+      // Should never reach here.
+      return -1;
+    }
+
+    /*
+      Return the number of times a given element is found in this List.
+
+      :arg x: An element to count.
+
+      :return: The number of times a given element is found in this List.
+      :rtype: `int`
+    */
+    proc count(x: eltType): int {
+      _enter();
+
+      var result = 0;
+
+      for i in 1.._size do {
+        if x == _getRef(i) then
+          result += 1;
+      }
+
+      _leave();
+      return result;
+    }
+
+    /*
+       Sort the elements of this List in place using their default sort order.
+
+       .. warning::
+
+        Sorting the contents of this List may invalidate existing references
+        to the elements contained in this List.
+    */
+    proc sort() {
+      // TODO: Need to look in the Sort module to see the API.
+    }
+
+    /*
+      Sort the items of this List in place using a comparator.
+
+      .. warning::
+
+        Sorting the elements of this List may invalidate existing references
+        to the elements contained in this List.
+
+      :arg comparator: A comparator object used to sort this List.
+    */
+    proc sort(comparator) {
+      // TODO: Need to look in the Sort module to see the API. 
+    }
+
+    /*
+      Index this List via subscript. Returns a reference to the element at a
+      given index in this List.
+
+      :arg i: The index of the element to access.
+
+      :return: An element from this List.
+
+      :throws IllegalArgumentError: If the given index is out of bounds.
+    */
+    proc const this(i: int) ref throws {
+      _enter();
+
+      //
```

Comment:
Looks like you might mean to delete the below code?

Reply:
I inserted that code after a long discussion with @vasslitvinov about whether the `this` method for `list` should throw, or behave similarly to `this` for arrays. What is your opinion on the matter?

Reply:
Oh, I see. I'd probably follow the array strategy for bounds checking (halt rather than throw).

Reply:
After some further thought I'm convinced halting is the right decision.